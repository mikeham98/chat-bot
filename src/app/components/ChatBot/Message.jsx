import React from 'react';
import PropTypes from 'prop-types';
import styles from '../../../themes/components/messages.scss';
import isPopulatedArray from "../../util/isPopulatedArray";
import Media from "./Media/Media";
import Options from "./Options/Options";
import {MediaObjectPropTypes, OptionsPropTypes} from "../../config/propTypes";

export default class Message extends React.PureComponent {
    constructor(props) {
        super(props);
        this.onClickOption = this.onClickOption.bind(this);
    }

    onClickOption(optionId) {
        const {id} = this.props;
        this.props.onClickOption(optionId, id)
    }

    showMedia() {
        return isPopulatedArray(this.props.media);
    }

    showOptions() {
        return isPopulatedArray(this.props.options);
    }

    render() {
        const {body, showDateTime, dateTime, media, options, onClickBody, selectedOption, color} = this.props;
        return (
            <div className={styles.messageContent}>
                <div className={styles.message} style={{backgroundColor: color}} onClick={onClickBody}>
                    <span className={styles.messageBody}>{body}</span>
                    {this.showOptions() && (
                        <Options
                            selectedOption={selectedOption}
                            options={options}
                            onClick={this.onClickOption}
                        />
                    )}
                </div>
                {this.showMedia() && (
                    <Media
                        media={media}
                    />
                )}
                {showDateTime && <span className={styles.messageCreatedAt}>{dateTime}</span>}
            </div>
        );
    }
}

Message.defaultProps = {
    color: ''
};

Message.propTypes = {
    onClickOption: PropTypes.func.isRequired,
    id: PropTypes.number,
    body: PropTypes.string,
    showDateTime: PropTypes.bool,
    dateTime: PropTypes.string,
    media: MediaObjectPropTypes,
    options: OptionsPropTypes,
    onClickBody: PropTypes.func.isRequired,
    selectedOption: PropTypes.number,
    color: PropTypes.string,
};