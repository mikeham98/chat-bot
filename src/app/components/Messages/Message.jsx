import React from 'react';
import styles from '../../../themes/messages.scss';

const Message = ({body, dateTime}) => {
    return (
        <div className={styles.message} data-tip={dateTime}>
            {body}
        </div>
    );
};

export default React.memo(Message);