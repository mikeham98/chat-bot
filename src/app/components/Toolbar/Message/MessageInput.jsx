import React from 'react';
import classnames from 'classnames';
import styles from '../../../../themes/toolbar.scss';

class MessageInput extends React.PureComponent {
    displayError() {
        const {error, touched} = this.props;
        return error && touched;
    };

    returnClassName() {
        const classNames = [styles.messageInputWrapper];
        if (this.displayError()) {
            classNames.push(styles.messageInputError);
        }
        return classnames(classNames);
    }

    render() {
        const {value, onChange, onBlur} = this.props;
        return (
            <input
                id={'message'}
                name={'message'}
                placeholder={'Type a message...'}
                type={'text'}
                value={value}
                onChange={onChange}
                onBlur={onBlur}
                className={this.returnClassName()}
            />
        );
    }
}

export default MessageInput;