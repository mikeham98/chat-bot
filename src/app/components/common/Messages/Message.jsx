import React from 'react';
import styles from '../../../../themes/messages.scss';

const Message = ({body}) => {
    return (
        <div className={styles.message}>
            {body}
        </div>
    );
};

export default React.memo(Message);