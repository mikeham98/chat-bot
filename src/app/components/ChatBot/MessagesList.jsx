import React from 'react';
import classnames from 'classnames';
import styles from '../../../themes/components/messages.scss';
import isPopulatedArray from "../../util/isPopulatedArray";
import Message from "./Message";
import {left, right} from "../../config/messages.config";
import formatDateTime from "../../util/formatDateTime";

export default class MessagesList extends React.PureComponent {
    constructor(props) {
        super(props);
        this.state = {
            clickedMessageId: null
        };
        this.messageContainerRef = this.messageContainerRef.bind(this);
    }

    componentDidMount() {
        this.scrollToBottom();
    }

    componentDidUpdate(prevProps) {
        if ((this.props.replying && !prevProps.replying) || (this.props.messages.length > prevProps.messages.length)) {
            this.scrollToBottom();
        }
    }

    scrollToBottom() {
        this.messageContainer.scrollTop = this.messageContainer.scrollHeight;
    }

    returnMessageClassName(position) {
        const classNames = [styles.messageWrapper];
        if (position === right) {
            classNames.push(styles.rightMessage);
        } else {
            classNames.push(styles.leftMessage);
        }
        return classnames(classNames);
    }

    returnMessages() {
        const {messages, currentUserId} = this.props;
        if (isPopulatedArray(messages)) {
            console.log(messages);
            return messages.map(message => {
                const position = currentUserId === message.createdBy.userId ? right : left;
                // showDateTime is implemented so that only one message at a time can show the date/time
                const showDateTime = message.id === this.state.clickedMessageId;
                return (
                    <div key={message.id}
                         onClick={() => this.setState({clickedMessageId: showDateTime ? null : message.id})}
                         className={this.returnMessageClassName(position)}>
                        <Message
                            showDateTime={showDateTime}
                            dateTime={formatDateTime(message.createdAt)}
                            body={message.content.body}
                        />
                    </div>
                );
            })
        }
    }

    messageContainerRef(ref) {
        this.messageContainer = ref;
    }

    render() {
        const {replying, botName} = this.props;
        return (
            <div
                ref={this.messageContainerRef}
                className={styles.messageContainer}
            >
                {this.returnMessages()}
                {replying && <span className={styles.replyInProgress}>{botName} is replying...</span>}
            </div>
        )
    }
}