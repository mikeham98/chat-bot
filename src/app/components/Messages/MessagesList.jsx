import React from 'react';
import moment from 'moment';
import classnames from 'classnames';
import styles from '../../../themes/messages.scss';
import isPopulatedArray from "../../util/isPopulatedArray";
import Message from "./Message";
import {left, right} from "../../config/messages.config";
import formatDateTime from "../../util/formatDateTime";

export default class MessagesList extends React.PureComponent {
    constructor(props) {
        super(props);
        this.state = {
            clickedMessageId: null
        };
        this.messageContainerRef = this.messageContainerRef.bind(this);
    }

    componentDidMount() {
        this.scrollToBottom();
    }

    componentDidUpdate(prevProps) {
        if ((this.props.replying && !prevProps.replying) || (this.props.messages.length > prevProps.messages.length)) {
            this.scrollToBottom();
        }
    }

    scrollToBottom() {
        this.messageContainer.scrollTop = this.messageContainer.scrollHeight;
    }

    returnMessageClassName(position) {
        const classNames = [styles.messageWrapper];
        if (position === right) {
            classNames.push(styles.rightMessage);
        } else {
            classNames.push(styles.leftMessage);
        }
        return classnames(classNames);
    }

    returnMessages() {
        const {messages, currentUserId} = this.props;
        if (isPopulatedArray(messages)) {
            return messages.map(message => {
                const position = currentUserId === message.createdBy.userId ? right : left;
                // showDateTime is implemented so that only one message at a time can show the date/time
                const showDateTime = message.id === this.state.clickedMessageId;
                return (
                    <div key={message.id} onClick={() => this.setState({clickedMessageId: showDateTime ? null : message.id})} className={this.returnMessageClassName(position)}>
                        <Message
                            showDateTime={showDateTime}
                            dateTime={formatDateTime(message.createdAt)}
                            body={message.body}
                        />
                    </div>
                );
            })
        }
    }

    messageContainerRef(ref) {
        this.messageContainer = ref;
    }

    render() {
        const {replying} = this.props;
        return (
            <div
                ref={this.messageContainerRef}
                className={styles.messageContainer}
            >
                {this.returnMessages()}
                {replying && <span className={styles.replyInProgress}>ChatBot is replying...</span>}
            </div>
        )
    }
}