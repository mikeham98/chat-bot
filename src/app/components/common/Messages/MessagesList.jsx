import React from 'react';
import moment from 'moment';
import classnames from 'classnames';
import styles from '../../../../themes/messages.scss';
import isPopulatedArray from "../../../util/isPopulatedArray";
import Message from "./Message";
import {left, right} from "../../../config/messages.config";

export default class MessagesList extends React.PureComponent {
    componentDidMount() {
        this.messageContainer.scrollTop = this.messageContainer.scrollHeight;
    }

    returnMessageClassName(position) {
        const classNames = [styles.messageWrapper];
        if(position === right) {
            classNames.push(styles.rightMessage);
        }else {
            classNames.push(styles.leftMessage);
        }
        return classnames(classNames);
    }

    returnMessages() {
        const {messages, currentUserId} = this.props;
        if (isPopulatedArray(messages)) {
            return messages.map(message => {
                const position = currentUserId === message.createdBy.userId ? right : left;
                return (
                    <div className={this.returnMessageClassName(position)}>
                        <Message
                            dateTime={moment(message.createdAt).fromNow()}
                            body={message.body}
                        />
                    </div>
                );
            })
        }
    }

    render() {
        return (
            <div
                ref={(ref) => this.messageContainer = ref}
                className={styles.messageContainer}
            >
                {this.returnMessages()}
            </div>
        )
    }
}