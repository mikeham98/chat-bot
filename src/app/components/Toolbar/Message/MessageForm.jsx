import React from 'react';
import {Formik} from 'formik';
import MessageInput from './MessageInput';
import IconButton from "../../common/Icon/Button";
import sendLogo from '../../../../../assets/icons/right-arrow-1.svg';
import styles from '../../../../themes/toolbar.scss';

export default class MessageForm extends React.Component {
    constructor(props) {
        super(props);
        this.handleSubmit = this.handleSubmit.bind(this);
    }

    handleSubmit({message}, {setSubmitting, resetForm}) {
        const callback = () => {
            setSubmitting(false);
            resetForm()
        };
        this.props.sendMessage(message, callback);
    }

    render() {
        return (
            <Formik
                initialValues={this.props.initialValues}
                onSubmit={this.handleSubmit}
                // validationSchema={Yup.object().shape({
                //     email: Yup.string()
                //         .email()
                //         .required('Required'),
                // })}
            >
                {formikProps => {
                    const {
                        values,
                        touched,
                        errors,
                        isSubmitting,
                        handleChange,
                        handleBlur,
                        handleSubmit,
                    } = formikProps;
                    return (
                        <form className={styles.messageFormWrapper} onSubmit={handleSubmit}>
                            <MessageInput
                                value={values.message}
                                onChange={handleChange}
                                onBlur={handleBlur}
                                touched={touched.message}
                                error={errors.message}
                            />
                            <IconButton
                                type='submit'
                                icon={sendLogo}
                                style={{margin: '0 10px'}}
                                color={values.message.length ? '#172B4D' : '#939aab'}
                                height='30px'
                                width='30px'
                                disabled={isSubmitting}
                            />
                        </form>
                    );
                }}
            </Formik>
        )
    }
}

MessageForm.defaultProps = {
    initialValues: {
        message: ''
    }
};