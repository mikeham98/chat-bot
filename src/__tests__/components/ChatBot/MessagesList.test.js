import React from 'react';
import {mount} from 'enzyme';
import MessagesList from '../../../app/components/ChatBot/MessagesList';
import Message from '../../../app/components/ChatBot/Message';

// formatDateTime has been mocked out due to the nature of re-running tests in the future and the current date changing.
jest.mock('../../../app/util/formatDateTime.js', () => ({
    __esModule: true,
    default: () => '10:30 am'
}));

describe('MessagesList', () => {
    let shallowedWrapper;
    let props = {};
    const wrapper = () => {
        if (!shallowedWrapper) {
            shallowedWrapper = mount(<MessagesList {...props}/>);
        }
        return shallowedWrapper;
    };

    beforeEach(() => {
        props = {
            messages,
            replying: false,
            currentUserId: 1,
            selectedOption: 2,
            onClickOption: jest.fn()
        };
        shallowedWrapper = undefined;
    });

    describe('render', () => {
        it('should return a list of messages', () => {
            expect(wrapper()).toMatchSnapshot();
        });
        it('should return no messages', () => {
            props.messages = [];
            expect(wrapper()).toMatchSnapshot();
        });
        it('should return ChatBot is replying... text', () => {
            expect(wrapper().find('.replyInProgress')).toHaveLength(0);
            wrapper().setProps({
                replying: true
            });
            expect(wrapper().find('.replyInProgress')).toHaveLength(1);
        });
    });

    describe('functionality', () => {
        describe('scrollToBottom', () => {
            beforeEach(() => {
                jest.spyOn(MessagesList.prototype, 'scrollToBottom');
            });

            afterEach(() => {
                jest.clearAllMocks();
            });

            it('should be called on mount', () => {
                expect(MessagesList.prototype.scrollToBottom).not.toHaveBeenCalled();
                // calling wrapper mounts the component
                wrapper();
                expect(MessagesList.prototype.scrollToBottom).toHaveBeenCalledTimes(1);
            });
            it('should not be called on update (unchanged props)', () => {
                expect(MessagesList.prototype.scrollToBottom).not.toHaveBeenCalled();
                wrapper();
                // clear the spyOn scrollToBottom which is called on mount
                jest.clearAllMocks();
                wrapper().update();
                expect(MessagesList.prototype.scrollToBottom).not.toHaveBeenCalled();
            });
            it('should not be called on update of random prop being passed in', () => {
                expect(MessagesList.prototype.scrollToBottom).not.toHaveBeenCalled();
                wrapper();
                // clear the spyOn scrollToBottom which is called on mount
                jest.clearAllMocks();
                // re-render with new prop
                wrapper().setProps({
                    foo: 'bar'
                });
                expect(MessagesList.prototype.scrollToBottom).not.toHaveBeenCalled();
            });
            it('should be called on change of replying prop from false to true', () => {
                expect(MessagesList.prototype.scrollToBottom).not.toHaveBeenCalled();
                wrapper();
                // clear the spyOn scrollToBottom which is called on mount
                jest.clearAllMocks();
                // re-render with new prop
                wrapper().setProps({
                    replying: true
                });
                expect(MessagesList.prototype.scrollToBottom).toHaveBeenCalledTimes(1);
            });
            it('should be called on change of messages prop increasing in length', () => {
                expect(MessagesList.prototype.scrollToBottom).not.toHaveBeenCalled();
                wrapper();
                // clear the spyOn scrollToBottom which is called on mount
                jest.clearAllMocks();
                // re-render with new prop
                wrapper().setProps({
                    messages: [
                        ...messages,
                        {
                            "id": 3,
                            "content": {
                                "body": "I am good thanks, how are you?"
                            },
                            "read": {
                                "status": false,
                                "timestamp": null
                            },
                            "recipient": {
                                "userId": 1,
                                "userName": "mikeham98"
                            },
                            "createdAt": "2019-07-23T12:02:30+00:00",
                            "createdBy": {
                                "userId": 100,
                                "userName": "johnSmith100"
                            }
                        }
                    ]
                });
                expect(MessagesList.prototype.scrollToBottom).toHaveBeenCalledTimes(1);
            });
        });
        describe('onClickOption', () => {
            it('should be called when calling onClickOption on Message', () => {
                expect(props.onClickOption).not.toHaveBeenCalled();
                wrapper().find(Message).at(0).props().onClickOption();
                expect(props.onClickOption).toHaveBeenCalledTimes(1);
            });
        });
        describe('showDateTime', () => {
            it('should be set to true for the first message', () => {
                expect(wrapper().find(Message).at(0).props().showDateTime).toBeFalsy();
                wrapper().find(Message).at(0).props().onClickBody();
                wrapper().update();
                expect(wrapper().find(Message).at(0).props().showDateTime).toBeTruthy();
            });
            it('should be set to true for the first message and then false on re-clicking', () => {
                expect(wrapper().find(Message).at(0).props().showDateTime).toBeFalsy();
                wrapper().find(Message).at(0).props().onClickBody();
                wrapper().update();
                expect(wrapper().find(Message).at(0).props().showDateTime).toBeTruthy();
                wrapper().find(Message).at(0).props().onClickBody();
                wrapper().update();
                expect(wrapper().find(Message).at(0).props().showDateTime).toBeFalsy();
            });
            it('should be set to true for the first message then false on click of the second' +
                ' message which now has showDateTime set to true', () => {
                expect(wrapper().find(Message).at(0).props().showDateTime).toBeFalsy();
                expect(wrapper().find(Message).at(1).props().showDateTime).toBeFalsy();
                wrapper().find(Message).at(0).props().onClickBody();
                wrapper().update();
                expect(wrapper().find(Message).at(0).props().showDateTime).toBeTruthy();
                expect(wrapper().find(Message).at(1).props().showDateTime).toBeFalsy();
                wrapper().find(Message).at(1).props().onClickBody();
                wrapper().update();
                expect(wrapper().find(Message).at(0).props().showDateTime).toBeFalsy();
                expect(wrapper().find(Message).at(1).props().showDateTime).toBeTruthy();
            });
        });
    });
});

const messages = [{
    "id": 0,
    "content": {
        "body": "Hello bot, what should I watch today",
    },
    "read": {
        "status": true,
        "timestamp": "2019-07-23T12:00:15+00:00"
    },
    "recipient": {
        "userId": 1,
        "userName": "mikeham98"
    },
    "createdAt": "2019-07-23T12:00:00+00:00",
    "createdBy": {
        "userId": 100,
        "userName": "johnSmith100"
    }
}, {
    "id": 1,
    "content": {
        "body": "Hello Mike, I have a few suggestions...",
        "options": [
            {
                "id": 1,
                "option": "Back To The Future"
            },
            {
                "id": 2,
                "option": "Stranger Things"
            }
        ]
    },
    "read": {
        "status": false,
        "timestamp": null
    },
    "recipient": {
        "userId": 100,
        "userName": "johnSmith100"
    },
    "createdAt": "2019-07-23T12:01:30+00:00",
    "createdBy": {
        "userId": 1,
        "userName": "mikeham98"
    },
    "selectedOption": 1
}];