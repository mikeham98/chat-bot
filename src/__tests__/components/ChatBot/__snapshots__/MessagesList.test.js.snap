// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessagesList render should return a list of 2 messages 1`] = `
<MessagesList
  botName="I am a bot"
  color="#1ccb9e"
  currentUserId={1}
  id={1}
  messages={
    Array [
      Object {
        "content": Object {
          "body": "Hello bot, what should I watch today",
        },
        "createdAt": "2019-07-23T12:00:00+00:00",
        "createdBy": Object {
          "userId": 100,
          "userName": "johnSmith100",
        },
        "id": 0,
        "read": Object {
          "status": true,
          "timestamp": "2019-07-23T12:00:15+00:00",
        },
        "recipient": Object {
          "userId": 1,
          "userName": "mikeham98",
        },
      },
      Object {
        "content": Object {
          "body": "Hello Mike, I have a few suggestions...",
          "options": Array [
            Object {
              "id": 1,
              "option": "Back To The Future",
            },
            Object {
              "id": 2,
              "option": "Stranger Things",
            },
          ],
        },
        "createdAt": "2019-07-23T12:01:30+00:00",
        "createdBy": Object {
          "userId": 1,
          "userName": "mikeham98",
        },
        "id": 1,
        "read": Object {
          "status": false,
          "timestamp": null,
        },
        "recipient": Object {
          "userId": 100,
          "userName": "johnSmith100",
        },
        "selectedOption": 1,
      },
    ]
  }
  onClickOption={[MockFunction]}
  replying={false}
>
  <div
    className="messageContainer"
  >
    <div
      className="messageWrapper leftMessage"
      key="0"
    >
      <Message
        body="Hello bot, what should I watch today"
        color=""
        dateTime="10:30 am"
        id={0}
        onClickBody={[Function]}
        onClickOption={[MockFunction]}
        showDateTime={false}
      >
        <div
          className="messageContent"
        >
          <div
            className="message"
            onClick={[Function]}
            style={
              Object {
                "backgroundColor": "",
              }
            }
          >
            <span
              className="messageBody"
            >
              Hello bot, what should I watch today
            </span>
          </div>
        </div>
      </Message>
    </div>
    <div
      className="messageWrapper rightMessage"
      key="1"
    >
      <Message
        body="Hello Mike, I have a few suggestions..."
        color="#1ccb9e"
        dateTime="10:30 am"
        id={1}
        onClickBody={[Function]}
        onClickOption={[MockFunction]}
        options={
          Array [
            Object {
              "id": 1,
              "option": "Back To The Future",
            },
            Object {
              "id": 2,
              "option": "Stranger Things",
            },
          ]
        }
        selectedOption={1}
        showDateTime={false}
      >
        <div
          className="messageContent"
        >
          <div
            className="message"
            onClick={[Function]}
            style={
              Object {
                "backgroundColor": "#1ccb9e",
              }
            }
          >
            <span
              className="messageBody"
            >
              Hello Mike, I have a few suggestions...
            </span>
            <Options
              onClick={[Function]}
              options={
                Array [
                  Object {
                    "id": 1,
                    "option": "Back To The Future",
                  },
                  Object {
                    "id": 2,
                    "option": "Stranger Things",
                  },
                ]
              }
              selectedOption={1}
            >
              <div
                className="messageOptionWrapper"
              >
                <Option
                  disabled={true}
                  key="1"
                  onClick={[Function]}
                  option="Back To The Future"
                  selected={true}
                >
                  <div
                    className="messageOption messageOptionSelected"
                    onClick={[Function]}
                  >
                    Back To The Future
                  </div>
                </Option>
                <Option
                  disabled={true}
                  key="2"
                  onClick={[Function]}
                  option="Stranger Things"
                  selected={false}
                >
                  <div
                    className="messageOption messageOptionDisabled"
                    onClick={[Function]}
                  >
                    Stranger Things
                  </div>
                </Option>
              </div>
            </Options>
          </div>
        </div>
      </Message>
    </div>
  </div>
</MessagesList>
`;

exports[`MessagesList render should return no messages 1`] = `
<MessagesList
  botName="I am a bot"
  color="#1ccb9e"
  currentUserId={1}
  id={1}
  messages={Array []}
  onClickOption={[MockFunction]}
  replying={false}
>
  <div
    className="messageContainer"
  />
</MessagesList>
`;
